src/commands/dev/edge/watch.rs

```diff
@@ -37,11 +38,19 @@ pub fn watch_for_changes(
         // this allows the server to route subsequent requests
         // to the proper script
         let uploaded = setup::upload(&mut target, &deploy_target, &user, session_token, verbose);
-        if let Ok(token) = uploaded {
-            *preview_token = token;
-        } else {
-            refresh_session_channel.send(Some(()))?;
-            break;
+        match uploaded {
+            Ok(token) => {
+                *preview_token = token;
+            }
+            Err(err) => {
+                if let Some(err) = err.downcast_ref::<setup::BadRequestError>() {
+                    // no need to refresh session here, bad request indicates problem with user code
+                    StdOut::warn(&err.0);
```

Comment:
Someone else should review this bit, I'm not sure if the cloudflare API could ever send back a 400 if the session needs to be refreshed. The rest of the PR looks good though :)

Reply:
In my quick testing, I didn't notice this causing any issues -- I'm gonna merge this, and if it winds up being problematic we can follow up with a bugfix PR